### PR TITLE
check PnetCDF version at compile time

### PR DESCRIPTION
**Pull request by wkliao**
When building in a cross-compile environment, AC_RUN_IFELSE may not work.
This PR replaces AC_RUN_IFELSE with AC_COMPILE_IFELSE.

### CODE REVIEW

**Comment by wkliao** [comment]
There is no need to save LIBS, as AC_COMPILE_IFELSE is to compile only.
Yes, I did test it against older PnetCDF versions.

**Comment by DennisHeimbigner** [comment]
I am confused. I thought that the eason that the program needs to be run is that
it is not possible to do version arithmetic at compile time. That is why
it is important to run the program as opposed to only compiling it.
You imply that the test is somehow being done at compile time. How
does that occur?

**Comment by wkliao** [comment]
C preprocessor arithmetic operations can compare integer values of defined constants at compile time. Try the C codes I added in this PR.

**Comment by DennisHeimbigner** [comment]
ok, that makes sense.

File: configure.ac [hunk 1]
```diff
@@ -1102,21 +1102,13 @@ if test "x$enable_pnetcdf" = xyes; then
...
-                   int minor = PNETCDF_VERSION_MINOR;
-                   int version = major*1000 + minor;
-                   int ok = (version >= (1*1000 + 6));
-                   return (ok?0:1);]])],
-                 [pnetcdf16=yes],
-                 [pnetcdf16=no])
-  CFLAGS="$SAVECFLAGS"
-  LIBS="$SAVELIBS"
   AC_MSG_CHECKING([Is libpnetcdf version 1.6.0 or later?])
+  AC_COMPILE_IFELSE([AC_LANG_PROGRAM([], [[
```
Comment (edhartnett) [inline configure.ac]: Is this also being checked in the cmake build files?

**Comment by wkliao** [comment]
I think so. Although cmake approach is different (it calls string() to extract the version info), it also does not need to compile or run any test program.

When cross-compiling, configure.ac should avoid running any test program, i.e. calling AC_RUN_IFELSE. I tried the master branch on Edison@NERSC and it failed with this error.
```
checking for ncmpi_create in -lpnetcdf... yes
configure: error: in `/global/homes/w/wkliao/NetCDF/GIT/master':
configure: error: cannot run test program while cross compiling
See `config.log' for more details
```

**Comment by DennisHeimbigner** [comment]
I think the situation is hopeless for cross compile. I think all of the AC_CHECK_FUNCTION
and similar calls all create and execute test programs because they can only do
the check when trying to load the actual library.,

**Comment by wkliao** [comment]
AC_CHECK_FUNC does not run any test programs. It internally calls AC_LINK_IFELSE, linking only which is sufficient to know whether a library contains a function. When cross compiling, it is OK to compile and link test programs, but not run. In most of the cases, there is no need to run. This is true for PnetCDF and I believe same for NetCDF.

**Comment by edhartnett** [comment]
@wkliao you are correct, there should be no cases in which the netCDF autotools build needs to run code during configure. Long ago we had many such tests, but I removed them. As you say there should be no reason for them. The autoconf macros are very clever at figuring things out just from compiling, without actually running, code.

**Comment by WardF** [comment]
Pulling into a combined PR request to set the stage for larger merges early next week.

**Pull request merged**